### PR TITLE
Tflite round

### PR DESCRIPTION
**Pull request by siju-samuel**
ROUND operator for TFLITE

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by siju-samuel** [comment]
@jdduke @achowdhery @aselle Could you please help me to review or assign a reviewer for this tflite op, Thanks in advance.

**Review by jdduke — DISMISSED** [review]
We should probably move ceil/floor/round into elementwise.cc, but that can be done separately. Looks good, thanks!

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (siju-samuel) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: i havnt tested performance yet, i will check it whether this have any significant improvement.

**Comment by siju-samuel** [comment]
@jdduke thanks for the review, Updated scalar test case and removed the conflicts due to gather_nd op merge. Could you please approve again? Thanks

**Review by jdduke — DISMISSED** [review]

**Comment by hgadig** [comment]
@siju-samuel  Please rebase to resolve the branch conflicts.

**Comment by siju-samuel** [comment]
@jdduke @hgadig sorry for the trouble, could you please approve again. I needed to rebase due to the cos op merge. Thanks for the support for review.

**Comment by hgadig** [comment]
> @jdduke @hgadig sorry for the trouble, could you please approve again. I needed to rebase due to the cos op merge. Thanks for the support for review.

Sure. @jdduke  is the one who needs to approve.

**Comment by siju-samuel** [comment]
@jdduke please check and approve again.. Again got rebased due to Where and Rank ops merge.

**Comment by siju-samuel** [comment]
@jdduke could you please approve this? The approval got removed after rebasing.

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@jdduke thanks for your review and approval, ELU & REVERSE_SEQUENCE got merged after your approval, so i again needed to rebase. 
could you please approve again.  TIA

**Comment by siju-samuel** [comment]
@jdduke can u please approve once again. I'm afraid some other merge will again create some conflicts.

**Comment by siju-samuel** [comment]
@jdduke thanks for your review and approval, ELU & REVERSE_SEQUENCE got merged after your approval, so i again needed to rebase.
could you please approve again. TIA

File: tensorflow/lite/testing/generate_examples.py
Comment (jdduke) [reply tensorflow/lite/testing/generate_examples.py]: Ping, can you add the [] case here as well?

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@jdduke kindly approve again. Added scalar. Thanks for pointing it out.

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@rthadur gentle ping to merge this PR, I'm afraid of new conflicts happening because of other merges. Thanks in advance.

**Comment by rthadur** [comment]
> @rthadur gentle ping to merge this PR, I'm afraid of new conflicts happening because of other merges. Thanks in advance.

@siju-samuel we can not merge these changes until conflicts are resolved , can you please rebase your branch once again.

**Comment by siju-samuel** [comment]
@jdduke gentle ping to approve again. Again got conflicts and rebased.
@rthadur The rebasing is done. Thank you for notifying.

**Comment by siju-samuel** [comment]
@jdduke Thanks for helping me previously with the review and approval. Gentle ping to approve again. Again got conflicts and rebased.

**Review by jdduke — CHANGES_REQUESTED** [review]
After digging into this some more, this implementation is incorrect. TensorFlow uses what's known as bankers rounding, which rounds half value to the nearest even. See also https://www.tensorflow.org/api_docs/python/tf/math/round. I don't think we can land this as-is.

**Comment by jdduke** [comment]
Hmm, it might be that the optimized version is correct, but I'm seeing slightly different results on mobile. It's possible TensorFlow(Mobile) has the same issue.

**Comment by siju-samuel** [comment]
@jdduke Thank you very much for your deep review. Re-based and fixed the round issue. Could you please check the current implementation.

**Comment by siju-samuel** [comment]
@jdduke Thank you very much for your deep review. Re-based and fixed the round issue. Could you please check the current implementation.

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: Don't you need FE_TONEAREST?

Comment (siju-samuel) [reply tensorflow/lite/kernels/internal/reference/reference_ops.h]: @jdduke I didnt expliitly set the mode assuming the mode will be FE_TONEAREST. I understand its better to set the mode and use. Now i updated the code to set the mode. Please help to review.

Comment (jdduke) [reply tensorflow/lite/kernels/internal/reference/reference_ops.h]: Shouldn't this be std::rint? I don't think we want a long.

File: tensorflow/lite/kernels/round_test.cc
Comment (jdduke) [inline tensorflow/lite/kernels/round_test.cc]: Can you add 1.5 as well?

**Review by jdduke — CHANGES_REQUESTED** [review]

**Comment by siju-samuel** [comment]
@jdduke Thank you for the reviews and helping me to improve this PR. I have added testcase and updated with rint instead of lrint.

**Comment by siju-samuel** [comment]
@jdduke Thank you for the reviews and helping me to improve this PR. I have added testcase and updated with rint instead of lrint.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Please alphabetize.

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: Shouldn't this be <fenv>?

**Review by jdduke — CHANGES_REQUESTED** [review]
Just a few comments, then I think we're ready.

**Comment by siju-samuel** [comment]
@jdduke Thank you so much for helping me find the missing things. The code is updated based on 2 review comments, Could you please check again and approve if everything is fine.

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@hgadig @rthadur Could you please help to merge this PR before any conflicts happen. TIA.

**Comment by siju-samuel** [comment]
@jdduke i needed to rebase due to MATRIX_DIAG op addition, could you please help to approve again. Thanks in advance.

**Review by jdduke — DISMISSED** [review]

**Comment by jdduke** [comment]
Thanks for your patience!

**Comment by siju-samuel** [comment]
@jdduke Sorry to bother you again, QUANTIZE operator got merged and i needed to rebase. Kindly go through again and approve.

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@hgadig @rthadur Gentle ping to merge this PR. TIA

**Comment by siju-samuel** [comment]
@jdduke  `make_round_test` function moved from   `generate_examples.py` to `generate_examples_lib.py`
due to commit `17ca6549ba Refactoring: Split logic from generate_examples py_binary target.`
Please check and approve. Thanks a lot for your support.

**Comment by jdduke** [comment]
Out of curiosity, is there a specific model you're using which requires this op? Is it blocking you?

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
Im just studying and exploring tflite with my own model. Im using my private build, not blocking me.

**Comment by siju-samuel** [comment]
@jdduke Because of `MATRIX_SET_DIAG` i rebased it recently, please check and approve once again. Thank you.

**Review by jdduke — APPROVED** [review]

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@rthadur Gentle ping, please help to merge this PR. TIA.

**Comment by jdduke** [comment]
Just a note, this is stuck on review internally as older Android NDK versions don't work properly with cfenv. In general, it requires min API 21, which is too restrictive (we still support API 18/19). Can you think of a workaround?

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: I don't think we need a temporary var here, just use i directly.

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: Add a comment that cfenv (for fesetround) is not yet supported universally on Android.

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: static_cast<int>(floor_val)

File: tensorflow/lite/kernels/internal/reference/reference_ops.h
Comment (jdduke) [inline tensorflow/lite/kernels/internal/reference/reference_ops.h]: floor_val + 1.0f

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@jdduke Thanks for the review, the code is updated accordingly. Gentle ping to check again.

**Comment by siju-samuel** [comment]
@jdduke Thanks for the review, the code is updated accordingly. Gentle ping to check again.

**Comment by siju-samuel** [comment]
@jdduke Could you please help to review again and approve if no issues. Thanks.

**Review by jdduke — DISMISSED** [review]

**Comment by siju-samuel** [comment]
@rthadur could you please help to finish the internal review and merge this PR. TIA

**Comment by siju-samuel** [comment]
@rthadur @hgadig could you please help to merge this PR? TIA. Its quite sometime since it got approved and im afraid of conflicts.

**Comment by jdduke** [comment]
Sorry for the delay, this hit another snag in our import pipeline. Working to land now.

**Comment by jdduke** [comment]
This has been merged in PR #24918.

**Comment by siju-samuel** [comment]
Thanks @jdduke

**Pull request closed without merging**
